python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +63,1034 @@
     from ._dataset_ref import DatasetRef
     from ._dataset_type import DatasetType
     from ._storage_class import StorageClass
-    from .dimensions import DataCoordinate
+    from .datastore.cache_manager import AbstractDatastoreCacheManager
 
     # Define a new special type for functions that take "entity"
     Entity: TypeAlias = DatasetType | DatasetRef | StorageClass | str
 
 
+class FileIntegrityError(RuntimeError):
+    """The file metadata is inconsistent with the metadata supplied by
+    the datastore.
+    """
+
+
+class FormatterNotImplementedError(NotImplementedError):
+    """Formatter does not implement the specific read or write method
+    that is being requested.
+    """
+
+
+class FormatterV2:
+    """Interface for reading and writing datasets using URIs.
+
+    The formatters are associated with a particular `StorageClass`.
+
+    Parameters
+    ----------
+    file_descriptor : `FileDescriptor`, optional
+        Identifies the file to read or write, and the associated storage
+        classes and parameter information.  Its value can be `None` if the
+        caller will never call `Formatter.read` or `Formatter.write`.
+    ref : `DatasetRef`
+        The dataset associated with this formatter. Should not be a component
+        dataset ref.
```

Comment:
Which storage class should this have?  A storage class appears at least here and (twice) in the `file_descriptor`, and there are many different potential storage classes and scenarios in play (read/write, component/full-dataset/[dis]assembly, in-formatter coercion, datastore-level coercion).

More than anything I'd like to drop support for any of those scenarios we don't need, but given that we've accumulated a lot historically and probably can't drop anything other than edge-case combinations without a deprecation period, we probably need to document all possible scenarios (of course, with the rules for those as simple as we can make them).

Reply:
As written it will have the original storage class that the user passed in if the formatter says it can accept that, else it will have the registry storage class. Before I implemented `can_accept` it was always going to be the registry storage class. Having this available could make the file descriptor `readStorageClass` obsolete. Will need to think.